docs/dapp/sapphire/browser.md

```diff
@@ -0,0 +1,188 @@
+ // to use when deploying to other networks.
+-const HARDHAT_NETWORK_ID = '1337';
++const HARDHAT_NETWORK_ID = '23295'; // Sapphire Testnet
+ 
+ // This is an error code that indicates that the user canceled a transaction
+ const ERROR_CODE_TX_REJECTED_BY_USER = 4001;
+@@ -225,14 +227,22 @@
+ 
+   async _initializeEthers() {
+     // We first initialize ethers by creating a provider using window.ethereum
+-    this._provider = new ethers.providers.Web3Provider(window.ethereum);
++    this._provider = sapphire.wrap(new ethers.providers.Web3Provider(window.ethereum));
+ 
+-    // Then, we initialize the contract using that provider and the token's
+-    // artifact. You can do this same thing with your contracts.
++    // Then, we initialize two contract instances:
++    // - _token: Used for eth_calls (e.g. balanceOf, name, symbol)
++    // - _tokenWrite: Used for on-chain transactions (e.g. transfer)
+     this._token = new ethers.Contract(
+       contractAddress.Token,
+       TokenArtifact.abi,
+-      this._provider.getSigner(0)
++      this._provider,
++    );
++    this._tokenWrite = new ethers.Contract(
++      contractAddress.Token,
++      TokenArtifact.abi,
++      sapphire.wrap(new ethers.providers.Web3Provider(window.ethereum).getSigner())
+     );
+   }
+ 
+@@ -294,7 +304,7 @@
+ 
+       // We send the transaction, and save its hash in the Dapp's state. This
+       // way we can indicate that we are waiting for it to be mined.
+-      const tx = await this._token.transfer(to, amount);
++      const tx = await this._tokenWrite.transfer(to, amount);
+       this.setState({ txBeingSent: tx.hash });
+ 
+       // We use .wait() to wait for the transaction to be mined. This method
+@@ -360,14 +370,14 @@
+       return true;
+     }
+ 
+-    this.setState({ 
+-      networkError: 'Please connect Metamask to Localhost:8545'
++    this.setState({
++      networkError: 'Please connect to Sapphire ParaTime Testnet'
+     });
+ 
+     return false;
+```
+
+Beside the obvious change to the chain ID and wrapping ethers.js objects with the
+Sapphire wrapper you probably noticed that we initialized **two** contract
+instances:
+
+- `this._token` object will be used for unsigned queries by Metamask and thus
+  no annoying popups will be shown to the user (this is the Hardhat way of
+  [setting `from` transaction field to all zeros][guide-transaction-calls]).
+  Queries which do not rely on the `msg.sender` field can be performed this 
+  way.
+  TODO: Are such transactions encrypted?
+- `this._tokenWrite` object will use Metamask to sign and encrypt the
+  `eth_call` query or the on-chain transaction (TODO: is this true?). The user will need to confirm
+  such action each time.
+
+### Trying it
+
+Start the frontend by typing `pnpm run start`. If all goes well the web server
+will spin up and your browser should automatically open
+`http://localhost:3000`.
+
+![Hardhat boilerplate frontend](../images/sapphire/hardhat-boilerplate-frontend1.png)
```

Comment:
Images could be optimized from 134KB to 29KB with commands from https://github.com/oasisprotocol/docs/commit/01a93e12cefa5cf09d2103a3e1a69f05d40068c8

Optimized:
[hardhat-boilerplate-frontend1.png](https://user-images.githubusercontent.com/3758846/199121768-23d8f9b3-f80b-44eb-8600-b14489add73e.png)
[hardhat-boilerplate-frontend2.png](https://user-images.githubusercontent.com/3758846/199121771-1f30f703-d62e-4f95-8a5f-7646be9abe67.png)
[hardhat-boilerplate-frontend3.png](https://user-images.githubusercontent.com/3758846/199121773-371a2356-5f8a-4e71-8e19-6fef4f46dcc1.png)

Reply:
Cool! FYI I needed to run `npx @squoosh/cli` to work.